kratos/spaces/ublas_space.h

```diff
@@ -734,8 +741,20 @@ class UblasSpace
             }
 
             if(empty) {
+            #ifdef KRATOS_DEBUG
+                // Check that Index Index term exists
+                for (std::size_t j = col_begin; j < col_end; ++j) {
+                    if (Acol_indices[j] == Index) {
+                        Avalues[j] = scale_factor;
+                        rb[Index] = 0.0;
+                        break;
+                    }
+                }
+                KRATOS_ERROR << "Diagonal term (" << Index << ", " << Index << ") is not defined in the system matrix" << std::endl;
+            #else
                 rA(Index, Index) = scale_factor;
```

Comment:
isnt this still some `find` operation?

Reply:
Yes, what alternative there is?. Only executed in debug.

Reply:
well, i would use the following (surely to be slightly adapted to make it to compile)

~~~cpp
auto pos = std::lower_bound(&AcolIndices[col_begin],&Acolndices[col_end]);
if pos != &Acolndices[col_end] //it was found
     rValues[pos - &AcolIndices[col_begin]] = scale_factor;
else
    KRATOS_ERROR ... 
~~~

this code shall be a bit faster than what you have, and it will use a binary search

if you do not like this, there is a function "find_element(i,i)" which will do the same as calling the lower bound.
i think my version should be slightly faster because we already have col_begin and col_end.

Reply:
Anycase looks like we have tests failing in full debug, hurray

Reply:
can you take a look to what are the cases failing? to know if it is a realistic scenario or just something we use in the tests (my bet is that there are constraints ... )

Reply:
Constraints should define diagonal as well. I am taking a look, in parallel other developments. And is ublas_space, each time I need to recompile half Kratos

Reply:
With your code: 

```
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolver
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithZeroContribution
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithConstraints
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithInactiveConstraints
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithConstraintsAuxiliarNode
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithConstraintsAuxiliarNodeInverted
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverAllDoFsMaster
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverAllDoFsMasterFromStructureSide
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementEliminationBuilderAndSolverWithZeroContribution
```

Reply:
my code was wrong. it should be 

~~~
pos - &Acolndices[0]
~~~

Reply:
Yes, still failing

Reply:
Okay now it only fails in Debug (in my machine) and the failing tests are the same:

```
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolver
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithZeroContribution
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithConstraints
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverWithInactiveConstraints
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementBlockBuilderAndSolverAllDoFsMaster
[  FAILED  ] KratosCoreFastSuite.BasicDisplacementEliminationBuilderAndSolverWithZeroContribution
```